### DISCUSSION TITLE
Additional PIN 2 HD Sense setting

### DISCUSSION THREAD

**Discussion by LittleGem** [keirf/flashfloppy#856] [Ideas]
Hi Keir,

I have just finished the design of a WD1772 floppy controller replacement which runs at either 8MHz and 16MHz (automatically switched by setting a config pin to either 0V or +5V) for both DD/SD and HD modes. I would like to use pin 2 of the gotek to control the density mode but the way it is programmed means that at idle (disk select not asserted) the output is always +5V whether pin2 = dens or pin = ndens is set in FF.CFG. I'm guessing that this is due to the output being open collector?

Observed behaviour is as follows:
With pin2 = dens, idle gives +5V. When the disk is selected DD = +5V, HD = 0V
With pin2 = ndens, idle gives +5V, When the disk is selected then DD = 0V, HD = +5V.

Could an additional setting be created for pin2 so that idle gives 0V and when the disk is selected then DD = 0V, HD = +5V?  

For the auto switching of the WD1772 clone  this means for drives without density signalling the default could be either DD/SD or HD depending on how the idle (gotek not selected) setting was configured. For drives that do have density signalling on pin 2 (due to modification) then the current settings work fine.

If you need further info please let me know. 

Regards,

Charlie

**Comment by keirf** [comment]
Hi Charlie,

Congratulations on another fine replacement part!

It's not particularly easy for FlashFloppy to do what you describe as the drive-select ISR has [tight timing requirements](https://github.com/keirf/flashfloppy/blob/master/src/gotek/floppy.c#L224). It is specialised to only assert lines on select, and deassert them on deselect. 

We could plausibly add a special secondary handler, as I do for Amiga_HD_ID() which toggles pin 34 on each deselect-select cycle. This could take responsibility for pin 2 at the expense of ISR latency.

For mixed multi-drive setups, it would be fairly easy to solder in a low-drop diode between the real drive's select line, and line 2, on the drive or even in the cable?

Regards,
Keir

**Reply by LittleGem** [reply]
Hi Keir,

Thanks for the prompt reply and kind words. 

Thinking through a bit more, it makes sense that pin 2 on a drive is open collector given that it is only asserted when selected and any drives that share pin 2 need to follow the same protocol. Therfore I have decided to change the logic sense on the VR177X (what I call the the WD1772/0 replacement) to match the normal FlashFloppy pin 2 dens setting.  If needed, I can provide the required interface for real drives (usually 0V DD/+5V HD) directly on the VR177X pcb with a suitable connector (this is really only relevant for the Atari ST(E)/MegaSTE at this time). 

BTW, as well determining the HD mode by way of sensing the DC level on the VR177X clock input, the VR177X can also determine the clock input frequency and change mode accordingly. With existing HD mods used in the Atari machines this would be the most likely use case. 

Regards,

Charlie
